pkg/epp/datastore/datastore.go

```diff
@@ -57,56 +61,94 @@ type Datastore interface {
-	PodGet(namespacedName types.NamespacedName) *PodMetrics
+	// PodGetAll returns all pods and metrics, including fresh and stale.
+	PodGetAll() []backendmetrics.PodMetrics
+	// PodList lists pods matching the given predicate.
+	PodList(func(backendmetrics.PodMetrics) bool) []backendmetrics.PodMetrics
+	PodUpdateOrAddIfNotExist(pod *corev1.Pod, pool *v1alpha2.InferencePool) bool
 	PodDelete(namespacedName types.NamespacedName)
-	PodResyncAll(ctx context.Context, ctrlClient client.Client)
-	PodGetAll() []*PodMetrics
-	PodDeleteAll() // This is only for testing.
-	PodRange(f func(key, value any) bool)
+	PodResyncAll(ctx context.Context, ctrlClient client.Client, pool *v1alpha2.InferencePool)
 
 	// Clears the store state, happens when the pool gets deleted.
 	Clear()
 }
 
-func NewDatastore() Datastore {
+func NewDatastore(parentCtx context.Context, pmf *backendmetrics.PodMetricsFactory) *datastore {
 	store := &datastore{
+		parentCtx:       parentCtx,
 		poolAndModelsMu: sync.RWMutex{},
 		models:          make(map[string]*v1alpha2.InferenceModel),
 		pods:            &sync.Map{},
+		pmf:             pmf,
 	}
 	return store
 }
 
 // Used for test only
-func NewFakeDatastore(pods []*PodMetrics, models []*v1alpha2.InferenceModel, pool *v1alpha2.InferencePool) Datastore {
-	store := NewDatastore()
-
-	for _, pod := range pods {
-		// Making a copy since in tests we may use the same global PodMetric across tests.
-		p := *pod
-		store.(*datastore).pods.Store(pod.NamespacedName, &p)
-	}
+func NewFakeDatastore(parentCtx context.Context, pmf *backendmetrics.PodMetricsFactory, pods []*backendmetrics.Pod, models []*v1alpha2.InferenceModel, pool *v1alpha2.InferencePool) Datastore {
+	store := NewDatastore(parentCtx, pmf)
 
 	for _, m := range models {
 		store.ModelSetIfOlder(m)
 	}
 
-	if pool != nil {
-		store.(*datastore).pool = pool
+	store.PoolSet(pool)
+
+	for _, pod := range pods {
+		// Making a copy since in tests we may use the same global PodMetric across tests.
+		p := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pod.NamespacedName.Name,
+				Namespace: pod.NamespacedName.Namespace,
+			},
+			Status: corev1.PodStatus{
+				PodIP: pod.Address,
+			},
+		}
+		store.PodUpdateOrAddIfNotExist(p, pool)
 	}
+
 	return store
 }
 
 type datastore struct {
+	// parentCtx controls the lifecycle of the background metrics goroutines that spawn up by the datastore.
+	parentCtx context.Context
 	// poolAndModelsMu is used to synchronize access to pool and the models map.
 	poolAndModelsMu sync.RWMutex
 	pool            *v1alpha2.InferencePool
 	// key: InferenceModel.Spec.ModelName, value: *InferenceModel
 	models map[string]*v1alpha2.InferenceModel
-	// key: types.NamespacedName, value: *PodMetrics
+	// key: types.NamespacedName, value: *backendmetrics.PodMetrics
 	pods *sync.Map
+	pmf  *backendmetrics.PodMetricsFactory
+}
+
+func (ds *datastore) flushPrometheusMetricsOnce(logger logr.Logger) {
```

Comment:
Can we move this to the file that invokes it, flushing prometheus metric is not a concern of the datastore.

Reply:
Also, why do we need this and PrintMetricsForDebugging? we can fuse both into one.

Reply:
You probably reviewed an older version? I think I already removed this from the datastore.

> we can fuse both into one.

Done. Was thinking they were dealing with different metrics (fetching backend metrics vs emitting epp metrics). However the "epp metrics" here is about translation of backend metrics as well, so having both in the `backend/metrics` package should be fine.

Reply:
You may not have pushed that change, flushPrometheusMetricsOnce still exists in this file.